rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,429 @@
+
+#### PIN and touch policies
+
+When we generate a new private key, we can provide it with a custom touch and PIN policy. These policies will be enforced whenever the private key is used for signing or decrypting, which may occur multiple times for each `tsh` command.
+
+The available PIN policies are:
+ - `never`: never prompt the user for PIN
+ - `always`: always prompt the user for PIN
+ - `once`: prompt the user for PIN just once per session, where a session is an open connection to the PIV module
+   - Due to the current keystore implementation reloading keys multiple times for every command, this may still lead to the user being prompted for PIN multiple times in each tsh call. This can potentially be reworked, but it will most likely not be fixed in the initial implementation.
+
+The available touch policies are:
+ - `never`: never prompt the user for touch
+ - `always`: always prompt the user for touch
+ - `cached`: cache user's touch for 15 seconds before prompting them for it again
+   - This should result in each `tsh` command requiring just one touch
+   - This could occur in the middle of a `tsh` command since we don't have control over when the touch is cached/expires.
+
+The user can provide `--piv-pin-policy=<never|always|once>` and `--piv-touch-policy=<never|always|cached>` to specify their desired policies.
+
+#### PIV slots
+
+PIV cards are required to have 4 primary slots, as well as the 20 retired key management slots. 
+
+Each slot has a different intended function and different default PIN and touch policies. Since we are manually setting the PIN and touch policies ourselves, we can technically use any of the 24 slots available in the same way.
+
+We will use slot `9a: PIV Authentication` by default, as it fits the teleport use case most closely. However, users will be able to provide their own desired slot with the flag `--piv-slot`, in case they are already using slot `9a` for a different login session or for a different PIV application, such as `yubikey-agent`.
+
+Note: Yubikeys also have an attestation slot, `f9`, which is not standardized by the PIV specifications.
+
+#### tsh PIV login session
+
+Usually, `tsh` checks the user's `~/.tsh/keys/proxy.example.com` for an existing public-private key pair. If found, that key pair will be used alongside any stored certificates to handle `tsh` calls.
+
+With PIV login session's we need a similar way to detect whether the user has an existing login session. In order to make PIV login a purely opt in feature, we will depend on information written to the user's tsh profile to determine whether the login session is PIV or non-PIV. Specifically, we will add a new `piv` field to the user's `~/.tsh/proxy.yaml` file which can also uniquely identify the PIV card and Slot being used for the login session.
```

Comment:
On second thought, using `proxy.yaml` in this way breaks the separation of concerns between `ProfileStatus` and `KeyStore`. Instead, it will probably make more sense to store this yaml data in the normal key file location - `~/.tsh/keys/proxy.example.com/user`. This way, we can determine whether we have an RSA private key, PIV private key, etc. by parsing the private key data in that file. This will also extend nicely to identity files. We can even PEM encode the PIV private key "data" with a unique type: `-----PIV PRIVATE KEY------`.

I'll hold off on updating the RFD for now to get other's thoughts.

Reply:
Yeah, some metadata file in the keys directory would probably be more suitable as this is not technically proxy settings.